File: src/test/java/personal/alexgarland/helloworldmvc/model/EmployeeTest.java
package personal.alexgarland.helloworldmvc.model;

import org.junit.Test;
import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertTrue;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

import static org.junit.Assert.assertFalse;


public class EmployeeTest {

	@Test
	public void returnsFullNameWithoutNickname() {
		Employee employee = new Employee(1, "Robert", "Dobalina");
		assertEquals("Robert Dobalina", employee.getFullName());
	}
	
	@Test
	public void returnsFullNameWithNickname() {
		Employee employee = new Employee(1, "Robert", "Dobalina", "Bob");
		assertEquals("Robert \"Bob\" Dobalina", employee.getFullName());
	}
	
	@Test
	public void returnsToString() {
		Employee employee = new Employee(1, "Robert", "Dobalina", "Bob");
		assertEquals("Employee 1: Robert \"Bob\" Dobalina", employee.toString());
	}
		
	@Test
	public void sortsAsExpected() {
		Employee employee1 = new Employee(1, "John", "Kennedy", "Jack");
		Employee employee2 = new Employee(2, "Robert", "Dobalina");
		Employee employee3 = new Employee(3, "James", "Fallon", "Jimmy");
		List<Employee> list = new ArrayList<Employee>();
		list.add(employee2);
		list.add(employee3);
		list.add(employee1);
		
		Collections.sort(list);
		
		assertTrue(list.get(0) == employee1);
		assertTrue(list.get(1) == employee2);
		assertTrue(list.get(2) == employee3);
	}
	
	@Test
	public void idAccessorsWork() {
		Employee employee = new Employee(1, "Robert", "Dobalina");
		employee.setId(2);
		assertEquals(2, employee.getId());
	}
	
	@Test
	public void firstNameAccessorsWork() {
		Employee employee = new Employee(1, "Bruce", "Jenner");
		employee.setFirstName("Caitlyn");
		assertEquals("Caitlyn", employee.getFirstName());
	}
	
	@Test
	public void lastNameAccessorsWork() {
		Employee employee = new Employee(1, "Kim", "Kardashian");
		employee.setLastName("Kardashian West");
		assertEquals("Kardashian West", employee.getLastName());
	}

	@Test
	public void nickNameAccessorsWork() {
		Employee employee = new Employee(1, "Kanye", "West", "The Louis Vuitton Don");
		employee.setNickName("The Lyor Cohen of Dior Homme");
		assertEquals("The Lyor Cohen of Dior Homme", employee.getNickName());
	}
	
	@Test
	public void copiesWithNewId() {
		Employee employee = new Employee(1, "Robert", "Dobalina");
		Employee newEmployee = employee.copyWithNewId(2);
		assertEquals(new Employee(2, "Robert", "Dobalina"), newEmployee);
	}
	
}

File: src/test/kotlin/personal/alexgarland/helloworldmvc/model/EmployeeTest.kt
package personal.alexgarland.helloworldmvc.model

import org.junit.Assert.assertEquals
import org.junit.Assert.assertTrue
import org.junit.Test
import java.util.*

class EmployeeTest {

    @Test
    fun returnsFullNameWithoutNickname() {
        val employee = Employee(1, "Robert", "Dobalina")
        assertEquals("Robert Dobalina", employee.fullName)
    }

    @Test
    fun returnsFullNameWithNickname() {
        val employee = Employee(1, "Robert", "Dobalina", "Bob")
        assertEquals("Robert \"Bob\" Dobalina", employee.fullName)
    }

    @Test
    fun returnsToString() {
        val employee = Employee(1, "Robert", "Dobalina", "Bob")
        assertEquals("Employee 1: Robert \"Bob\" Dobalina", employee.toString())
    }

    @Test
    fun sortsAsExpected() {
        val employee1 = Employee(1, "John", "Kennedy", "Jack")
        val employee2 = Employee(2, "Robert", "Dobalina")
        val employee3 = Employee(3, "James", "Fallon", "Jimmy")
        val list = ArrayList<Employee>()
        list.add(employee2)
        list.add(employee3)
        list.add(employee1)

        list.sort()

        assertTrue(list[0] === employee1)
        assertTrue(list[1] === employee2)
        assertTrue(list[2] === employee3)
    }

    @Test
    fun idAccessorsWork() {
        val employee = Employee(1, "Robert", "Dobalina")
        employee.id = 2
        assertEquals(2, employee.id.toLong())
    }

    @Test
    fun firstNameAccessorsWork() {
        val employee = Employee(1, "Bruce", "Jenner")
        employee.firstName = "Caitlyn"
        assertEquals("Caitlyn", employee.firstName)
    }

    @Test
    fun lastNameAccessorsWork() {
        val employee = Employee(1, "Kim", "Kardashian")
        employee.lastName = "Kardashian West"
        assertEquals("Kardashian West", employee.lastName)
    }

    @Test
    fun nickNameAccessorsWork() {
        val employee = Employee(1, "Kanye", "West", "The Louis Vuitton Don")
        employee.nickName = "The Lyor Cohen of Dior Homme"
        assertEquals("The Lyor Cohen of Dior Homme", employee.nickName)
    }

    @Test
    fun copiesWithNewId() {
        val employee = Employee(1, "Robert", "Dobalina")
        val newEmployee = employee.copyWithNewId(2)
        assertEquals(Employee(2, "Robert", "Dobalina"), newEmployee)
    }

}
